tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/Cache.scala

```diff
@@ -0,0 +1,177 @@
+package at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.caches
+
+import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
+import at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.Snapshottable
+
+import scala.collection.immutable.HashMap
+
+/**
+ * An abstract cache that assigns `TargetT`-typed values to `SourceT` typed values, while potentially modifying a
+ * `ContextT`-typed context.
+ *
+ * Example: IntValueCache assigns [[ArenaCell]]s to integers, while modifying a [[PureArena]].
+ *
+ * Some caches maintain a certain contract with the SMT solver, that is, they may add constraints to a
+ * [[at.forsyte.apalache.tla.bmcmt.smt.SolverContext]] on demand.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+abstract class Cache[ContextT, SourceT, TargetT] extends Snapshottable {
+
+  /* While it extends Snapshottable, there's no need to implement an actual snapshot/stack data structure. By assigning each
+   * cache entry a level, effectively a timestamp for when it was added, we can proxy the state of the stack, by filtering
+   * out entries, based on their levels.
+   *
+   * Caches need snapshot functionality, because reverting SMT contexts should revert arenas as well, thus freeing certain
+   * cells. Otherwise, we could end up with a cache pointing to a cell, which does not exist in a post-revert arena.
+   */
+
+  type LevelT = Int
+
+  /**
+   * A stack level, see [[Snapshottable]]
+   */
+  private var _level: LevelT = 0
+
+  // the base map, tracking the level at which each entry was added.
+  private var _cache: Map[SourceT, (TargetT, LevelT)] = HashMap()
+  // inherited classes can read, but not modify without going thorough the public interface methods
+  protected def cache: Map[SourceT, (TargetT, LevelT)] = _cache
+
+  // reverse mapping
+  private var _reverseCache: Map[TargetT, (SourceT, LevelT)] = HashMap()
+  // inherited classes can read, but not modify without going thorough the public interface methods
+  protected def reverseCache: Map[TargetT, (SourceT, LevelT)] = _reverseCache
+
+  def values(): Iterable[TargetT] = _cache.values.map(_._1)
```

Comment:
Hmm, you're probably right that `reverseCache.keys` is strictly better, it was this way originally and since it's not technically wrong it didn't stand out to me, but I'll change it to your suggestion.